src/shaka_scheme/system/base/Environment.hpp

```diff
@@ -0,0 +1,102 @@
+//
+// Created by dylan on 9/18/17.
+//
+
+#ifndef SHAKA_SCHEME_ENVIRONMENT_H
+#define SHAKA_SCHEME_ENVIRONMENT_H
+
+#include <map>
+#include "shaka_scheme/system/base/IEnvironment.hpp"
+#include "shaka_scheme/system/base/DataPair.hpp"
+#include "shaka_scheme/system/base/Symbol.hpp"
+
+namespace shaka {
+/**
+ * @brief Representation for a Scheme Environment
+ */
+class Environment : public IEnvironment<shaka::Symbol, NodePtr> {
+
+public:
+  using Key = shaka::Symbol;
+  using Value = NodePtr;
+
+  /**
+  * @brief Constructor for Environment
+  * @param parent Pointer to Parent Environment, nullptr if top environment
+  */
+  Environment(std::shared_ptr<IEnvironment<Key, Value>> parent);
+
+  /**
+   * @brief Destroys Environment
+   */
+  ~Environment() override;
+
+  /**
+   * @brief Gets pointer to the parent environment
+   * @return The pointer to the parent environment
+   */
+
+  std::shared_ptr<IEnvironment<Key, Value>> get_parent() override;
+
+  /**
+   * @brief Sets the Parent Environment to 'e'
+   * @param e Pointer to an Environment
+   */
+  void set_parent(std::shared_ptr<IEnvironment<Key, Value>> e);
+
+  /**
+   * @brief Sets an element in the Environment referred to by the 'key'
+   *        and sets its value to 'data'
+   * @param key They key to be used for lookup in the environment
+   * @param data The data associated with the key
+   */
+  void set_value(const Key& key, Value data) override;
+
+  /**
+   * @brief Returns a reference to the value referred to by the key. Key
+   *        must exist, otherwise, it throws a runtime error.
+   * @param key The key to lookup and return the value for.
+   * @return The value associated with the given key
```

Comment:
There should be another `@throws InvalidInputException when the key does not exist in the environment bindings`. The function should throw an `InvalidInputException`.